Source/WebCore/style/StyleTreeResolver.cpp

```diff
@@ -160,15 +160,20 @@ std::unique_ptr<RenderStyle> TreeResolver::styleForStyleable(const Styleable& st
 
 static void resetStyleForNonRenderedDescendants(Element& current)
 {
-    for (auto& child : childrenOfType<Element>(current)) {
+    auto descendants = descendantsOfType<Element>(current);
+    for (auto it = descendants.begin(); it != descendants.end();) {
+        auto& child = *it;
         if (child.needsStyleRecalc()) {
             child.resetComputedStyle();
             child.resetStyleRelations();
             child.setHasValidStyle();
         }
 
-        if (child.childNeedsStyleRecalc())
-            resetStyleForNonRenderedDescendants(child);
+        if (child.childNeedsStyleRecalc()) {
+            it.traverseNext();
+            child.clearChildNeedsStyleRecalc();
+        } else
+            it.traverseNextSkippingChildren();
```

Comment:
You're already iterating an explicit list of all descendants. Jumping around based on traverseNext() and traverseNextSkippingChildren(), rather than iterating the list items in order, is at least confusing, and maybe actually incorrect, or at least not having the efficient outcome you had hoped?

For the sake of size in the descendants list, and clarifying this iteration logic, it may be best to create a list of all _children_, and then appending to it as needed.